x/gamm/pool-models/balancer/pool.go

```diff
@@ -618,14 +618,27 @@ func (p Pool) SpotPrice(ctx sdk.Context, baseAsset, quoteAsset string) (sdk.Dec,
 		return sdk.Dec{}, errors.New("pool is misconfigured, got 0 weight")
 	}
 
+	defer func() {
+		// defer function to escape the panic when spot price overflows
+		if r := recover(); r != nil {
+			spotPrice = sdk.Dec{}
+			err = errors.New("spot price overflow")
+		}
+	}()
+
 	// spot_price = (Base_supply / Weight_base) / (Quote_supply / Weight_quote)
 	// spot_price = (weight_quote / weight_base) * (base_supply / quote_supply)
 	invWeightRatio := quote.Weight.ToDec().Quo(base.Weight.ToDec())
 	supplyRatio := base.Token.Amount.ToDec().Quo(quote.Token.Amount.ToDec())
 	fullRatio := supplyRatio.Mul(invWeightRatio)
 	// we want to round this to `SigFigs` of precision
-	ratio := osmomath.SigFigRound(fullRatio, types.SigFigs)
-	return ratio, nil
+	spotPrice = osmomath.SigFigRound(fullRatio, types.SigFigs)
+	if spotPrice.GT(sdk.NewDec(2).Power(160)) {
+		spotPrice = sdk.Dec{}
+		err = errors.New("spot price overflow")
```

Comment:
err, the prior error type was good! We just need two errors:
`ErrSpotPriceInternal` and `ErrSpotPriceOverflow`

Reply:
Ah I see what you meant now – should be fixed now!